### PR TITLE
Support for ray tracing with Embree via double-down

### PR DESCRIPTION
**Pull request by pshriwise**
## Description
This adds an experimental feature to DAGMC which allows Embree to be used as back-end ray tracer for DAGMC via the new (and also somewhat experimental) [double-down](https://github.com/pshriwise/double-down) project.

I'm creating a draft here for comments on the approach, to check it on CI, and for others to try it out if they are interested.

## Motivation and Context
Provides significant serial and parallel efficiency.

## Changes
Creates a type alias for the `moab::GeomQueryTool` object. The double down project implements a very similar interface.

### CODE REVIEW

**Comment by gonuke** [comment]
I think I follow what's going on and the concept seems OK to me.  While it's a little tedious to change it everywhere, I wonder if we want to use a name other than `GQT` so that we don't assume that its a `GeometryQueryTool` instance.

Also not sure why all the new scripts in `CI/docker`?

**Comment by pshriwise** [comment]
Good point, @gonuke. I'm not sure why all those updates are there. I should've checked the diff more carefully I suppose. I'll see what I can do about them. 

I'm happy to update the `GQT` name to something new. Any suggestions? If sticking with our acronym scheme, I was thinking maybe `RT` (`RayTracer`) or `RTI` (`RayTracingInterface`). In a vacuum, I'd probably go with `ray_tracer`.

Update:
Looks like some of the CI script relocations weren't tracked in the rebase I did. I'll take care of those updates manually.

**Comment by gonuke** [comment]
I'm OK with `ray_tracer`.  The current syntax standard is motivated by just how long the class names are, but I think `ray_tracer` is short enough.

**Comment by gonuke** [comment]
Where will `RTI.hpp` and `MOABRay.hpp` live?

**Comment by pshriwise** [comment]
Those are both part of double-down and will be provided as part of the `dd` target when finding the double down installation in CMake.

**Comment by pshriwise** [comment]
I'm going to upgrade this to a real PR. It's passing checks now and it's had a once-over from @gonuke (thanks!)

File: CI/travis/build_double_down.sh
Comment (gonuke) [inline CI/travis/build_double_down.sh:1]: Should we put this into some/all of our docker images?  I don't know how long it takes to build this, but it would be great it we didn't add to the build time for DAGMC for every PR.

**Review by gonuke — COMMENTED** [review]
It all looks good to me - just one strategic question about CI

File: CI/travis/build_double_down.sh
Comment (pshriwise) [reply CI/travis/build_double_down.sh:1]: Good point! I'll get that added to the image.

Comment (gonuke) [reply CI/travis/build_double_down.sh:1]: You may want to confer with @bam241 about how best to add it to the matrix within Travis or wait until he gets CircleCI going?

Comment (pshriwise) [reply CI/travis/build_double_down.sh:1]: Indeed. @bam241 what do you think?

Comment (bam241) [reply CI/travis/build_double_down.sh:1]: Either way works for me.

it will probably be easier for you to wait for the CircleCI to be merge.
(It is almost ready. I think if it got some more reviews it could be merged today)

**Comment by pshriwise** [comment]
Ok so after some discussion it looks like this PR will now depend on #692 where we can incorporate the Embree and double-down build in our CI images.

**Comment by gonuke** [comment]
Yup - and #692 is merged now.  Let me know if you need any help weaving the double down into the new CI/docker infrastructure

**Comment by bam241** [comment]
@pshriwise let me know if you need a hand with the new CI/docker infrastructure :)

**Comment by pshriwise** [comment]
Ok so it looks like there are a few failing builds on the `develop` branch in CI. In Travis, these builds were allowed failures. I did a little searching but didn't find anything about allowed failures for jobs in Circle. Is that possible @bam241?

I ask because any builds with Double Down should probably be allowed failures as well.

**Comment by pshriwise** [comment]
Ran a build in CI with the double down turned on for commit [8051e9b](https://app.circleci.com/pipelines/github/pshriwise/DAGMC/20/workflows/9db663a3-74c7-495d-b634-cf411a2731c8), but I've moved it into the merge workflow now as it's an experimental build. I'm assuming we don't want it to hold up future PR's if it fails.

**Comment by pshriwise** [comment]
One more note, I added a flag to do a dry run of all the docker builds without pushing them. I found this useful in making sure I wasn't partially replacing the CI images.

File: CI/Dockerfile_4_embree
Comment (gonuke) [inline CI/Dockerfile_4_embree]: I think this file is now unused/superfluous?

File: CI/travis/install.sh
Comment (gonuke) [inline CI/travis/install.sh:23]: Isn't double down already installed always now?

File: .circleci/config.yml
Comment (gonuke) [inline .circleci/config.yml]: Won’t this always report failure?  It ignores the previous status, which I assume prevents the test from reporting failure.  However, it doesn’t use the previous status to decide whether to report this failure.  Do we need to add something to the conditional?

**Review by gonuke — CHANGES_REQUESTED** [review]
I think this looks ok except that it may report a constantly failing status for DD

**Comment by gonuke** [comment]
I see the check mark next to 8051e9b, but I can’t find a report on CircleCI to see what it reports.  Maybe we should merge this with it in the PR workflow, and then another PR to move it to the merge workflow?

File: CI/Dockerfile_4_embree
Comment (pshriwise) [reply CI/Dockerfile_4_embree]: This whole file is, yes. Meant to remove it after my first attempt at adding this to CI. Thanks!

File: CI/travis/install.sh
Comment (pshriwise) [reply CI/travis/install.sh:23]: Embree is always present in our docker files, but double down will be built each time as we're building using it's `develop` branch. It's build time is very short relative to MOAB & DAGMC.

File: .circleci/config.yml
Comment (pshriwise) [reply .circleci/config.yml]: That whole block isn't really useful. I'll remove it.

Based on previous failed attempts to get that build working, that block will always report the same thing but the build will still fail. The `ignore_previous_status` line just ensures that this step in the job always runs and the `echo` command always returning a status of 0 should always result in this step passing -- even if the `echo`'d message doesn't make a lot of sense.

File: CI/travis/install.sh
Comment (gonuke) [reply CI/travis/install.sh:23]: Got it.  Yes, I forgot this was now on merge, rather than on PR.

File: .circleci/config.yml
Comment (pshriwise) [reply .circleci/config.yml]: https://github.com/svalinn/DAGMC/pull/693#issuecomment-667475806

Still curious about this @bam241. Will we be able to have a set of builds on merge with allowed failures that won't cause our `develop` branch show as failing CI?

**Comment by pshriwise** [comment]
> I see the check mark next to 8051e9b, but I can’t find a report on CircleCI to see what it reports. Maybe we should merge this with it in the PR workflow, and then another PR to move it to the merge workflow?

Yeah that build seems to be viewable only on my CircleCI account (as opposed to Svalinn) for some reason?? Not sure why. I moved the build back to the PR section for now so it should be viewable. Then I can make one final commit after approval moving it back to our set of merge jobs.

File: .circleci/config.yml
Comment (gonuke) [inline .circleci/config.yml]: This forces a build of MOAB `master` branch rather than using an image with a prebuilt MOAB.  Is that intentional?

File: .circleci/config.yml
Comment (gonuke) [reply .circleci/config.yml]: Not sure there is a clean way to allow failures in Circle right now

File: .circleci/config.yml
Comment (pshriwise) [reply .circleci/config.yml]: It is. There are changes to `RayHistory` in master that are necessary for double down.

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
It's running the test in CI (???) - do you need the filters as well?

**Comment by gonuke** [comment]
Now a test fails!?!?

**Comment by pshriwise** [comment]
Works with a restart. But it was a legitimate test failure, not just a timeout. :disappointed:  That'll have to be addressed in another PR.

**Comment by gonuke** [comment]
Merging before the test fails again :) - thanks @pshriwise

**Pull request merged**
